meter/goodwe-wifi.go

```diff
@@ -0,0 +1,172 @@
+package meter
+
+import (
+	"encoding/binary"
+	"net"
+	"time"
+
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/util"
+)
+
+type goodweServer struct {
+	conn      *net.UDPConn
+	inverters map[string]goodweInverter
+}
+
+type goodweInverter struct {
+	IP           string
+	pvPower      float64
+	netPower     float64
+	batteryPower float64
+	soc          float64
+}
+
+var server *goodweServer
+
+type goodWeWiFiMeter struct {
+	usage string
+	URI   string
+}
+
+func init() {
+	registry.Add("goodwe-wifi", NewGoodWeWifiFromConfig)
+}
+
+//go:generate go run ../cmd/tools/decorate.go -f decorateGoodWeWifi -b *goodWeWiFiMeter -r api.Meter -t "api.Battery,Soc,func() (float64, error)"
+
+func NewGoodWeWifiFromConfig(other map[string]interface{}) (api.Meter, error) {
+	cc := struct {
+		capacity   `mapstructure:",squash"`
+		URI, Usage string
+		Cache      time.Duration
+	}{
+		Cache: time.Second,
+	}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	return NewGoodWeWiFiMeter(cc.URI, cc.Usage)
+}
+
+func NewGoodWeWiFiMeter(uri string, usage string) (api.Meter, error) {
+	meter := &goodWeWiFiMeter{
+		usage: usage,
+		URI:   uri,
+	}
+
+	server, err := NewServer()
+	if err != nil {
+		return nil, err
+	}
+	server.addInverter(uri)
+
+	// decorate api.BatterySoc
+	var batterySoc func() (float64, error)
+	if usage == "battery" {
+		batterySoc = meter.batterySoc
+	}
+
+	return decorateGoodWeWifi(meter, batterySoc), nil
+}
+
+func (m *goodWeWiFiMeter) CurrentPower() (float64, error) {
+	switch m.usage {
+	case "grid":
+		return server.inverters[m.URI].netPower, nil
+	case "pv":
+		return server.inverters[m.URI].pvPower, nil
+	case "battery":
+		return server.inverters[m.URI].batteryPower, nil
+	}
+	return 0, api.ErrNotAvailable
+}
+
+func (m *goodWeWiFiMeter) batterySoc() (float64, error) {
+	return server.inverters[m.URI].soc, nil
+}
+
+func NewServer() (*goodweServer, error) {
+	if server == nil {
```

Comment:
Dito bzgl. parallelen Zugriffen: das wäre gut in ein `sync.Once` oder Mutex zu verpacken.